plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/listener/QueryInsightsListener.java

```diff
@@ -69,11 +80,49 @@ public QueryInsightsListener(final ClusterService clusterService, final QueryIns
                 v -> this.queryInsightsService.getTopQueriesService(MetricType.LATENCY).setWindowSize(v),
                 v -> this.queryInsightsService.getTopQueriesService(MetricType.LATENCY).validateWindowSize(v)
             );
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(TOP_N_CPU_QUERIES_ENABLED, v -> this.setEnableTopQueries(MetricType.CPU, v));
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(
+                TOP_N_CPU_QUERIES_SIZE,
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.CPU).setTopNSize(v),
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.CPU).validateTopNSize(v)
+            );
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(
+                TOP_N_CPU_QUERIES_WINDOW_SIZE,
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.CPU).setWindowSize(v),
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.CPU).validateWindowSize(v)
+            );
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(TOP_N_MEMORY_QUERIES_ENABLED, v -> this.setEnableTopQueries(MetricType.MEMORY, v));
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(
+                TOP_N_MEMORY_QUERIES_SIZE,
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.MEMORY).setTopNSize(v),
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.MEMORY).validateTopNSize(v)
+            );
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(
+                TOP_N_MEMORY_QUERIES_WINDOW_SIZE,
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.MEMORY).setWindowSize(v),
+                v -> this.queryInsightsService.getTopQueriesService(MetricType.MEMORY).validateWindowSize(v)
```

Comment:
We should delegate these methods to QueryInsightsService, instead of exposing the topQueryService here. For example, in QueryInsightsService, there is a method:

```
public void validateWindowSize(MetricType type, int windowSize) {
  //delegate further if there is TopQueriesService base class, else resolve the metricType here
}
```

Reply:
Otherwise QueryInsightsListener might as well hold direct references to the TopQueriesService objects!

Reply:
Good point. Let me refacor this part to make it more maintainable.